# Python Day 2!

### What will we cover?
* Some more lists
* For and while loops
* Dictionaries!
* If-else statements (let's play rock-paper-scissors against the computer)
* Functions

## Lists

- A list is a collection of Python objects, such as strings or integers.

In [ ]:
# Create a list of three strings

animals = ['dog', 'cat', 'mouse']

print(animals)

### Methods

- "append" is a method of the list object "animals".

- It knows how to attach things to the end of the list.

In [ ]:
# Attach something to the end of the list

animals.append('rat')
print(animals)

- "sort" is also a method of the list animals. It puts the things in the list in order.

In [ ]:
# Sort a list in order

animals.sort()

print(animals)

- You can make an empty list.

In [ ]:
anotherList = []

print(anotherList)

## Exercise

- Create a list containing three integers. Be sure the integers *are not* in numerical order.
- Sort the list and print it out.
- Using a for loop, print out each integer in your list, one per line.

## List indexes

- Indexes are used to access individual items in a list.
- Indexes are integers, starting from 0.

In [ ]:
animals = ['dog', 'cat', 'mouse', 'rat']

print('animals: ', animals)

print('animals[0]:', animals[0])

print('animals[3]:', animals[3])

## Slices

- You can specify a range of indexes -- this is called a "slice".

- The slice is specified as a starting index, a colon, and a number that is **one more than the ending index** you want.

In [ ]:
print('animals[0:2]:', animals[0:2])

- You can omit one or both indexes
  - The starting index defaults to 0.
  - The ending index defaults to the end of the list.

In [ ]:
# Use the default starting index.

print('animals[:2]:', animals[:2])

# Use the default ending index.

print('animals[1:]:', animals[1:])

# Use both defaults.

print('animals[:]: ', animals[:])

- Negative indexes start from the end

In [ ]:
print('animals[-1]:', animals[-1])

- Guido van Rossum (inventor of Python), *Why Python uses 0-based indexing.* October 24, 2013:
  - http://python-history.blogspot.com/2013/10/why-python-uses-0-based-indexing.html
  
  >Let's first look at use cases. Probably the most common use cases for slicing are "get the first n items" and "get the next n items starting at i".
  
  > Using 0-based indexing, half-open intervals, and suitable defaults (as Python ended up having), they are beautiful: a[:n] and a[i:i+n].
  

## Exercise 09

- Using a slice, print "cat" and "mouse" from the list of animals.

In [ ]:
animals = ['dog', 'cat', 'mouse', 'rat']
# your code here:

## For and While loops

How do you loop over the elements of a list?

In [ ]:
animals = ['cat', 'dog', 'monkey']
# loop over the animals:

In [ ]:
# While loop is another useful iteration methods

i = 1
while i < 6: # when this is no longer true, break out!
  print(i)
  i += 1

In [ ]:
brain_power_left = 10
while brain_power_left != 0:
  print('study moreeee')
  brain_power_left = brain_power_left - 1 # losing brain power

print("ok you can go to sleep now.")

# "if" statements

- An "if" statement causes the indented code to do something under certain conditions.

In [ ]:
x = 1

# Test if x is equal to 1.  The double equal sign tests for equality.

if x == 1:
    print('x equals 1')

x equals 1


- "else" covers the other conditions.

In [ ]:
x = 77

if x == 1:
    print('x equals 1')
else:
    print('x is not equal to 1')

x is not equal to 1


- You can also have one or more "elif" statements to cover other conditions. 

In [ ]:
x = 2

if x == 1:
    print('x equals 1')
elif x == 2:
    print('x equals 2')
else:
    print('x is not equal to 1 or 2')

x equals 2


In [ ]:
# Same as above, but now x is set to 3.

x = 3

if x == 1:
    print('x equals 1')
elif x == 2:
    print('x equals 2')
else:
    print('x is not equal to 1 or 2')

x is not equal to 1 or 2


- The "in" operator tests if an element is in a collection

In [ ]:
# Check if something is in a list

animals = ['dog', 'cat', 'mouse']

if 'dog' in animals:
    print('dog is in animals')
else:
    print('dog is not in animals')

dog is in animals


- "not in" tests if an elment is **not** in a collection

In [ ]:
# Check if something is not in a list

if 'rat' not in animals:
    animals.append('rat')
    
print(animals)

['dog', 'cat', 'mouse', 'rat']


- Every object in Python is either True or False, when used in an "if" statement.
  - A list with at least one item is True.  An empty list is False.
  - The same for a string.
  - The same for any other kind of collection.

In [ ]:
mylist = []
if mylist:
    print('mylist is not empty')
else:
    print('mylist is empty')

mylist is empty


## Let's play rock-paper-scissors against the computer!!

Some tools you'll need:

In [9]:
# You can ask the user for an input!

user_input = input('What day is it today? ')

if user_input == 'Tuesday':
  print('Correct!')
else:
  print('Fix your clock...')

What day is it today? idk
Check your clock :)


Can we make the code above more flexible? Let's say I write "I think it's Tuesday but I'm not sure", how can we count that as a correct answer?

In [ ]:
user_input = input('What day is it today? ')

if 'Tuesday' in user_input: # use "in" operator
  print('Correct!')
else:
  print('Check your clock lol')

### Random module

In [ ]:
# You can use the 'random' module to come up with random numbers! 

import random

foods = ['pasta', 'pizza', 'burgers']
user_input = input('My dearest human, I made you pasta, pizza, and burgers today. What would you like to eat? ')

# Randomly choose a food!
selected_food = random.choice(foods)
if selected_food == user_input:
  print(f"At your command. Enjoy your {selected_food}.")
else:
  print(f"Just kiddin, there's only {selected_food} on the menu today.")

In [ ]:
### CODE FOR ROCK-PAPER-SCISSORS



# Dictionaries

- A dictionary is a collection of key/value pairs.
- A dictionary is like a list, but:
  - The indexes (that is, the keys) are not necessarily integers.
  - The elements are not stored in any particular order.

In [ ]:
species = {'dog': 'Canis familiaris',
           'cat': 'Felis catus',
           'mouse': 'Mus musculus'}

print(species)

- Dictionaries values are accessed by keys.

In [ ]:
print(species['cat'])

- To add an item (a key/value pair) to a dictionary, simply assign a value to the new key.

In [ ]:
species['rat'] = 'Rattus norvegicus'

print(species)

### Lists vs. dictionaries

- Use a list when the order matters:

|Index|Value|
|---|---|
| 0 | 'dog' |
| 1 | 'cat' |
| 2 | 'mouse' |

- Use a dictionary when you want to look up values at random (quickly):

|Key|Value|
|---|---|
| 'dog' | 'Canis familiaris' |
| 'cat' | 'Felis catus' |
| 'mouse' | 'Mus musculus' |


### Brackets and braces

- Lists are created with square brackets:

In [ ]:
animals = ['dog', 'cat', 'mouse']

- Dictionaries are created with curly braces:

In [ ]:
species = {'dog': 'Canis familiaris', 'cat': 'Felis catus', 'mouse': 'Mus musculus'}

- But indexing is always done with square brackets:

In [ ]:
print(animals[0])
print(species['dog'])

### Dictionary keys and values

In [ ]:
# Get all the keys using the keys method

print(species.keys())

dict_keys(['dog', 'cat', 'mouse'])


In [ ]:
# Get all the values using the values method

print(species.values())

dict_values(['Canis familiaris', 'Felis catus', 'Mus musculus'])


### Iterating over a dictionary

- When you use a dictionary in a for loop, you get each key.

In [ ]:
# Iterate over a dictionary

for animal in species:
    print(animal, ': ', species[animal])

In [ ]:
d = {'person': 2, 'cat': 4, 'spider': 8}

for animal, legs in d.items(): # another way to iterate
    print('A {} has {} legs'.format(animal, legs))

### Exercise: build and iterate through a dictionary 

For every element in num, square it and build a dictionary mapping the element to its squared value. 

In [ ]:
nums = [0, 1, 2, 3, 4]
# your code here:


# Functions

- A function is used to package some code in a way that is easy to test and re-use.
- You have already used functions.
  - Some functions, like "print" are built-in functions.
  - But you can make your own functions.

In [ ]:
# Function example:

def celsiusToFahrenheit(temp_in_c):
    temp_in_f = (9 / 5) * temp_in_c + 32
    return temp_in_f

In [ ]:
c = 100.0
f = celsiusToFahrenheit(c)
print(c, 'C is', f, 'F') 

c = 0.0
f = celsiusToFahrenheit(c)
print(c, 'C is', f, 'F')

c = -40.0
f = celsiusToFahrenheit(c)
print(c, 'C is', f, 'F')

100.0 C is 212.0 F
0.0 C is 32.0 F
-40.0 C is -40.0 F


## Exercise 18

- Write a function that converts Fahrenheit to Celsius
- Print a Fahrenheit to Celsius conversion table from 32 F to 212 F
- The forumla is:
  - C = 5/9 x (F - 32)
  
- Hints:
  - You will need a for loop using the range function.
  - Parentheses effect the order of arithmetic operations.

In [ ]:
# your code here:


In [ ]:
# dictionary mapping car model to number of cars on the road in the U.S.
carsOnRoad = {'ACURA': 27458,
'ALFA ROMEO': 3735,
'AUDI': 34839,
'BMW': 50956,
'BUICK': 35521,
'CADILLAC': 23297,
'CHEVROLET': 330381,
'CHRYSLER': 13857,
'DODGE': 43756,
'FIAT': 1339,
'FORD': 411035,
'GENESIS': 3585,
'GMC': 101759,
'HONDA': 266044,
'HYUNDAI': 139378,
'INFINITI': 16097,
'JAGUAR': 4320,
'JEEP': 176713,
'KIA': 125302,
'LAND ROVER': 14500,
'LEXUS': 50458,
'LINCOLN': 21280,
'MAZDA': 61199,
'MERCEDES': 69505,
'MINI': 5288,
'MITSUBISHI': 12197,
'NISSAN': 161317,
'PORSCHE': 12193,
'RAM': 127682,
'SMART': 15023,
'SUBARU': 136519,
'TESLA': 36800,
'TOYOTA': 347571,
'VOLKSWAGEN': 69933,
'VOLVO': 23770}

In [ ]:
# write a function to compute the average number of cars on the road
# this is challenging!

## Rock-paper-scissors but better!

We're going to play rock-paper-scissors again but this time, we will use functions (which makes our program much more efficient and reproducible).

In [ ]:
### Code here:
